File: leetcode/99/P99.rb
# 99. Recover Binary Search Tree
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def recover_tree(root)
	$node1 = $node2 = $prev = nil
	search root
	val = $node1.val
	$node1.val = $node2.val
	$node2.val = val
end

def search node
	return if node == nil
	search node.left
	if $prev != nil
		if node.val < $prev.val
			$node1 = $prev if $node1 == nil
			$node2 = node
		end
	end
	$prev = node
	search node.right
end

File: leetcode/99/src/P99.java
/**
 * Created by qwordy on 2/22/16.
 * 99. Recover Binary Search Tree
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public void recoverTree(TreeNode root) {
		search(root);
		int val = node1.val;
		node1.val = node2.val;
		node2.val = val;
	}

	void search(TreeNode node) {
		if (node == null)
			return;
		search(node.left);
		if (prev != null) {
			if (node.val < prev.val) {
				if (node1 == null) node1 = prev;
				node2 = node;
			}
		}
		prev = node;
		search(node.right);
	}

	TreeNode node1, node2, prev;
}

public class P99 {
}
